File: graph/code/K站中转内最便宜的航班.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        if (src == dst) {
            return 0;
        }

        int[] previous = new int[n];
        int[] current = new int[n];
        for (int i = 0; i < n; i++) {
            previous[i] = Integer.MAX_VALUE;
            current[i] = Integer.MAX_VALUE;
        }
        previous[src] = 0;
        for (int m = 0; m <= k; m++) {
            current[src] = 0;
            for (int i = 0; i < flights.length; i++) {
                int cur = flights[i][1];
                int pre = flights[i][0];
                int dest = flights[i][2];

                if (previous[pre] != Integer.MAX_VALUE)
                    // 注意如果题目这里没有说最多经过k次，那么这里可以改成 current[cur] = Math.min(current[cur], current[pre] + dest);
                    current[cur] = Math.min(current[cur], previous[pre] + dest); 
            }
            previous = current.clone();
        }
        return current[dst] == Integer.MAX_VALUE ? -1 : current[dst];
    }
}

File: datastructure/07graph/code/K站中转内最便宜的航班.cpp
class Solution {
public:
    int max_num = 100000000;
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        vector<vector<int>> dp (k + 2, vector<int>(n, max_num));
        dp[0][src] = 0;

        for (int i = 1; i <= k + 1; i++) {
            for (auto &flight : flights) {
                dp[i][flight[1]] = min(dp[i][flight[1]], dp[i - 1][flight[0]] + flight[2]);
            }
        }

        int ans = max_num;

        for (int i = 1; i <= k + 1; i++) {
            ans = min(ans, dp[i][dst]);
        }

        return ans < max_num ? ans : -1;
    }
};